File: src/m2vscan.c
/*
 * m2vscan.c
 *
 * Author: Tomi Ollila too ät iki fi
 *
 *	Copyright (c) 2004 Tomi Ollila
 *	    All rights reserved
 *
 * Created: Sun Sep 26 13:23:29 EEST 2004 too
 * Last modified: Thu 28 Nov 2013 22:01:22 +0200 too
 *
 * This program is licensed under the GPL v2. See file COPYING for details.
 */

#ifndef _POSIX_SOURCE
#define _POSIX_SOURCE
#endif

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

#define DBGS 0
#include "x.h"
#include "zzob.h"
#include "bufwrite.h"

#include "ghdrs/m2vscan_priv.h"

#define BB {
#define BE }

#if 1
#define ASSERT_EQ(a, b) do { if ((a) != (b)) \
  xerrf("Assertion failed: " #a "(%d) != " #b "(%d)\n", (a), (b)); } while (0)
#else
#define ASSERT_EQ(a, b) do { } while (0)
#endif

int main(int argc, char ** argv)
{
  char * ifile;
  char * ofile;
  int ifd, ofd;
  struct stat st;

  if (argc != 3)
    xerrf("Usage: %s ifile ofile\n", argv[0]);

  ifile = argv[1];
  ofile = argv[2];

  if ((ifd = open(ifile, O_RDONLY)) < 0)
    xerrf("open '%s' as O_RDONLY failed:", ifile);

  if (fstat(ifd, &st) < 0)
    xerrf("fstat '%s' failed:", ifile);

  if ((ofd = open(ofile, O_WRONLY|O_CREAT|O_TRUNC, 0644)) < 0)
    xerrf("open '%s', (for write) failed:", ofile);

  s__scan(ifd, ofd, st.st_size);
  return 0;
}

static void s__scan(int ifd, int ofd,
		    off_t fsize) /* protoadd GCCATTR_NORETURN */
{
  ZeroZeroOneBuf zzob;
  unsigned char ibuf[65536];

  FILE * ofh = fdopen(ofd, "w");

  int pictures = 0;
  char pictypes[1025];
  int gops = -1;
  int goppic = 0;
  off_t seqpos = 0;
  int asr = 0;
  int maxwidth = 0, maxheight = 0;

  /* should put all other vars to their structs too */
  struct {
    int hour, minute, second, frame, closed, time;
  } gop = { 0, 0, 0, 0, 0, 0 };
  int printgopinfo = 0;
  int discpicadj = 0;

  int prevpc = 0;

  if (ofh == NULL)
    xerrf("fdopen \"w\" failed:");

  zzob_init(&zzob, ifd, ibuf, sizeof ibuf, -1);

  while (zzob_data(&zzob, false))
    {
      int cpc = zzob.pos * 100 / fsize;

      if (cpc != prevpc)
	{
	  printf("\r- Scanning video at %jd of %jd (%d%%) ",
		 (intmax_t)zzob.pos, (intmax_t)fsize, cpc);
	  fflush(stdout);
	  prevpc = cpc;
	}

      if (zzob.len >= 3)
	{
	  unsigned char * data = zzob.data; /*printf("%d\n", data[3]); */

	  switch (data[3])
	    {
	    case 0xb3:	/* sequence header */
	      seqpos = zzob.pos;
	      if (zzob.len > 7)
		{
		  int frame_rate = data[7] & 0x0f;
		  int width = (data[4] << 4) + (data[5] >> 4);
		  int height = ((data[5] & 0x0f) << 8) + data[6];

		  if (frame_rate != 3)
		    {
		      xerrf("\rCurrently we can only support PAL (25fps) "
			    "frame rate.\n");
		    }
		  asr = data[7] >> 4;
		  if (width > maxwidth) maxwidth = width;
		  if (height > maxheight) maxheight = height;

		}
	      break;

	    case 0xb8: /* group of pictures (GOP) header */
	      if (printgopinfo) {
		/* XXX currently a problem */
		fprintf(stderr,"\nZero picture gop at filepos %jd..., ",
			(intmax_t)zzob.pos);
		exit(1);
	      }

	      if (zzob.len > 7)
		{
		  /* int drop_frame_flag = data[4] & 0x80; */
		  gop.hour   = (data[4] & 0x7f) >> 2;
		  gop.minute = ((data[4] & 0x03) << 4) + (data[5] >> 4);
		  gop.second = ((data[5] & 0x07) << 3) + (data[6] >> 5);
		  gop.frame  = ((data[6] & 0x1f) << 1) + (data[7] >> 7);
		  gop.closed = data[7] & 0x40;
		  /* int broken_gop_flag = data[7] & 0x20; */

		  gop.time = (((gop.hour * 60 + gop.minute)
			       * 60 + gop.second) * 25) + gop.frame;
		  if (pictures) {
		    pictypes[pictures - goppic] = '\0';
		    fprintf(ofh, "%3d  %s\n", pictures - goppic, pictypes);
		    memset(pictypes, '_', sizeof pictypes);
		  }
		  else
		    fprintf(ofh, "#fileoffset  gop  frame ts asr _ time _   "
			    "c/o  # of frames ver2\n");

		  if (gop.time != pictures - discpicadj)
		    {
		      fprintf(stderr,"\nWarning: timecode discontinuity at "
			      "filepos %jd ", (intmax_t)zzob.pos);
		      fprintf(stderr, " %02d:%02d:%02d.%02d -- ",
			      gop.hour, gop.minute, gop.second, gop.frame);
		      fprintf(stderr, "%d %d\n", gop.time, pictures);
		      discpicadj = pictures - gop.time;
		    }
		  printgopinfo = 1;

		  gops++;
		}
	      break;
	    case 0x00: /* picture header */
	      if (zzob.len > 7)
		{
		  int temperal_seq_number = (data[4] << 2) + ((data[5] & 0xc0) >> 6);
		  int frame_type = ((data[5] & 0x38) >> 3);

		  pictypes[temperal_seq_number] = "_IPBDXXX"[frame_type];

		  if (printgopinfo)
		    {
		      char c;
		      if (frame_type == 1 && temperal_seq_number < 10)
			c = temperal_seq_number + '0';
		      else
			c = '-';
		      goppic = pictures;
		      fprintf(ofh, "%10jd %5d %6d %c  %d  %02d:%02d:%02d.%02d  %c ",
			      (intmax_t)seqpos, gops, goppic, c, asr,
			      gop.hour, gop.minute, gop.second, gop.frame,
			      gop.closed? 'c': 'o');
		    }
		  printgopinfo = 0;
		}
	      pictures++;
	      break;
	    }
	}
    }

  long abr = pictures? (long)(zzob.pos / (pictures / 25) * 8): 0;

  pictypes[pictures - goppic] = '\0';
  fprintf(ofh, "%3d  %s\n", pictures - goppic, pictypes);
  fprintf(ofh, "end: size %jd frames %d gops %d abr %ld  w %d h %d\n",
	  (intmax_t)zzob.pos, pictures, ++gops, abr, maxwidth, maxheight);
  printf("\r- Scanning video at %jd of %jd (100%%).\n",
	 (intmax_t)zzob.pos, (intmax_t)fsize);
  exit(0);
}


/*
 * Local variables:
 * mode: c
 * c-file-style: "gnu"
 * tab-width: 8
 * End:
 */

File: unused/m2vscan.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Author: Tomi Ollila -- too ät iki fi
#
#	Copyright (c) 2004 Tomi Ollila
#	    All rights reserved
#
# Created: Wed Sep 01 21:15:17 EEST 2004 too
# Last modified: Thu Oct 07 23:42:09 EEST 2004 too
#
# This program is released under GNU GPL. Check
# http://www.fsf.org/licenses/licenses.html
# to get your own copy of the GPL license.

# This code waits for cleanup, after proof-of-concept stage has been passed.

import sys
import struct

class m2vbuf:
    SIZE = 65536

    def __init__(self, filename):
        self.file = open(filename, 'rb')
        self.buf = ''
        self.offset = 0
        #self.pos = 0
        self.len = 0
        self.gone = 0

        while True:
            self.fillbuf(self.len - 0)
            try:
                self.pos = self.buf.index('\000\000\001', 0, self.len)
                return
            except ValueError:
                pass

    def fillbuf(self, rest):
        if rest > 0:
            nr = self.SIZE - rest
            if nr == 0:
                raise ValueError
            self.buf = self.buf[self.offset:] + self.file.read(nr)
            self.gone += self.offset
        else:
            self.gone += self.len
            self.buf = self.file.read(self.SIZE)

        self.len = len(self.buf)
        self.offset = 0

        if self.len == rest:
            raise EOFError

    def filepos(self):
        return self.gone + self.offset

    def next(self):
        self.offset = self.pos
        while True:
            try:
                self.pos = self.buf.index('\000\000\001', \
                                          self.offset + 3, self.len)
                return self.buf[self.offset:self.pos]
            except ValueError:
                pass

            try:
                self.fillbuf(self.len - self.offset)
            except EOFError:
                if self.len == self.offset:
                    raise EOFError
                return self.buf[self.offset:]

from optparse import OptionParser

parser = OptionParser(usage='usage: %prog [-v] input-file output-file')

parser.add_option('-v', '--verbose', action='store_true', dest='verbose',
                  default=False, help='Verbose mode')

(options, args) = parser.parse_args()

if len(args) != 2:
    parser.print_usage()
    print 'input-file and/or output-file missing.'
    sys.exit(1)

def f_picture(G, buf):	G.pictures += 1
def f_user_data(G, buf):	pass
def f_seq_error(G, buf):	pass
def f_ext_start(G, buf):	pass
def f_seq_end(G, buf):	pass


def f_seq_header(G, buf):
    G.seqpos = G.f.filepos()
    bytes = struct.unpack('BBBBBBBBB', buf[4:13])

    horiz_size = (bytes[0] << 4) + (bytes[1] >> 4)
    vert_size = ((bytes[1] & 0x0f) << 8) + bytes[2]
    aspect_ratio = bytes[3] >> 4
    frame_rate = bytes[3] & 0x0f
    bit_rate = (bytes[4] << 10) + (bytes[5] << 2) + (bytes[6] >> 6)
    vbv_buffer_size = (bytes[6] << 5) + (bytes[7] >> 3)
    #constrained_parameter =
    #load_intra_quantizier_matrix=

    #print horiz_size, 'x', vert_size
    #print aspect_ratio, frame_rate, bit_rate

    if frame_rate != 3:
        print >> sys.stderr, \
              "Currently we can only support PAL (25fps) frame rate."

def f_gop(G, buf):
    print >> G.ofile, G.infoline, '%02d' % (G.pictures - G.goppic)
    bytes = struct.unpack('BBBB', buf[4:8])

    drop_frame_flag = bytes[0] & 0x80
    hour   = (bytes[0] & 0x7f) >> 2
    minute = ((bytes[0] & 0x03) << 4) + (bytes[1] >> 4)
    second = ((bytes[1] & 0x07) << 3) + (bytes[2] >> 5)
    frame  = ((bytes[2] & 0x1f) << 1) + (bytes[3] >> 7)
    closed_gop_flag = bytes[3] & 0x40
    broken_gop_flag = bytes[3] & 0x20

    # XXX below hardcoded PAL frame rate, 25 fps.
    time = (((hour * 60 + minute) * 60 + second) * 25) + frame
    if time != G.pictures:
        print 'Timecode problem at', G.f.filepos(), '...'
        print '%02d:%02d:%02d.%02d' % (hour, minute, second, frame)
        print time, G.pictures
        sys.exit(1)

    if closed_gop_flag:
        cg = 'c'
    else:
        cg = 'o'

    #print '%02d:%02d:%02d.%02d' % (hour, minute, second, frame)
    #print drop_frame_flag, closed_gop_flag, broken_gop_flag

    G.goppic = G.pictures
    G.infoline = '%10d %5d %6d  %02d:%02d:%02d.%02d  %s ' \
                   % (G.seqpos, G.gops, G.goppic,
                      hour, minute, second, frame, cg)
    G.gops += 1

#    G.infoline = '-- ' + str(G.seqpos) +' '+ str(G.gops) +' '+ str(G.goppic) + ' %02d:%02d:%02d.%02d' % (hour, minute, second, frame)



scode_actions = {
    0x00: f_picture,
    0xb2: f_user_data,
    0xb3: f_seq_header,
    0xb4: f_seq_error,
    0xb5: f_ext_start,
    0xb7: f_seq_end,
    0xb8: f_gop
    }


class G:
    pictures = 0
    seqpos = -1
    goppic = 0
    gops = 0
    infoline = '#file offset gop# frame#  _ time _   c/o  # of frames'

G.f = f = m2vbuf(args[0])
G.ofile = open(args[1], 'w')

while True:
    try:
        buf = f.next()
    except EOFError:
        print >> G.ofile, G.infoline, G.pictures - G.goppic
        print >> G.ofile, 'end: size', f.filepos(), \
              'frames', G.pictures, 'gops', G.gops, \
              'abr', f.filepos() / (G.pictures / 25) * 8
        sys.exit(0)

    a = ord(buf[3])

    try:
        scode_actions[a](G, buf)
    except KeyError:
        if a < 0xa0:
            # slice, ignoring for now.
            pass
        else:
            print >> sys.stderr, "File contains mpeg header 0x%02x." % a
            print >> sys.stderr, "Did you demultiplex your source?";
            sys.exit(1)

#    if G.gops == 20:
#        sys.exit(0)
